### ISSUE TITLE
Add User Role

### ISSUE THREAD

**Issue by MichaelMaherKamel** [nicoalbanese/kirimase#33]
It would be nice to add the user role as an enum, and it should reflect in the NextAuth Sessions and the DB

**Comment by pjholmes** [comment]
I agree this would be nice. Using Prisma, I was able to add this pretty easily. However, when I moved to Drizzle, I have not been able to make it work for days. Unfortunately, the  PgDrizzleAdapter.getUserByAccount will not include the User role even though the query does not limit the columns read. Perhaps because I made it an Enum, but not sure. I would much prefer for @nicoalbanese to do this right.

**Comment by MichaelMaherKamel** [comment]
Thank you @pjholmes 

Well, for me, I took another path, it might not be very practical, but it did the job for me till next-auth fixes it or adds a way for Drizzle like they did for Prisma as explained here -- https://authjs.dev/guides/basics/role-based-access-control

I have extended the User Schema in the auth.ts file to hold the user Role as below:

src > lib > db > schema > auth.ts
____________________________________

export const users = mysqlTable('user', {
  id: varchar('id', { length: 255 }).notNull().primaryKey(),
  name: varchar('name', { length: 255 }),
  email: varchar('email', { length: 255 }).notNull(),
  emailVerified: timestamp('emailVerified', {
    mode: 'date',
    fsp: 3,
  }).defaultNow(),
  image: varchar('image', { length: 255 }),
  role: mysqlEnum('role', ['admin', 'user']).notNull().default('user'),
})

In my protected page, which is /admin, I am fetching the server session and saving the session.user.id in a variable, then fetching this user from the DB and comparing the roles, for example

src > app > admin > page.tsx
__________________________________

import { getServerSession } from 'next-auth'
import { authOptions } from '../api/auth/[...nextauth]/route'

import { db } from '@/lib/db'
import { eq } from 'drizzle-orm'
import { users } from '@/lib/db/schema/auth'

import SignIn from '@/components/auth/SignIn'

const AdminPage = async () => {
  const session = await getServerSession(authOptions)
  if (session) {
    const userId = session.user.id
    const [u] = await db.select().from(users).where(eq(users.id, userId))
    return (
      <>
        Signed in as {session.user?.name}
        {u.role === 'admin' ? (
          <div>
            <br />
            <h1>Admin Page</h1>
          </div>
        ) : (
          <div>
            <h1>Not an Admin</h1>
          </div>
        )}
      </>
    )
  }
  return (
    <>
      Not signed in <br />
      <SignIn />
    </>
  )
}

export default AdminPage

========================

Cheers 🤞🏻

**Comment by MaarcusRenieroL** [comment]
You can add the following code if you are working with Prisma

In the `next-auth.d.ts` file

`ts
// @typescript-eslint/no-unused-vars
import type { Session, User } from "next-auth";
import type { JWT } from "next-auth/jwt";

declare module "next-auth/jwt" {
  interface JWT {
    id: string;
    role: enum;
  }
}

declare module "next-auth" {
  interface Session {
    user: User & {
      id: string;
      role: enum;
    };
  }
}
`

In your Middleware File 

`ts
import { getToken } from "next-auth/jwt";
import { withAuth } from "next-auth/middleware";
import { NextResponse } from "next/server";
import { type UserRole } from "@prisma/client";

const redirectBasedOnRole = (role: UserRole, pathname: string) => {
  if (pathname === "/") {
    const roleBasedUrl = `/${role.toLowerCase().replace("_", "-")}`;
    return roleBasedUrl;
  }
};

export default withAuth(
  async function middleware(req) {
    const pathname = req.nextUrl.pathname;

    const token = await getToken({ req });
    const isAuth = !!token;

    const unprotectedRoutes = ["/login", "/register"];
    const isApiRoute = req.nextUrl.pathname.startsWith("/api");

    if (!isAuth && !unprotectedRoutes.includes(pathname) && !isApiRoute) {
      return NextResponse.redirect(new URL("/login", req.url));
    }

    if (["/login", "/signup"].includes(req.nextUrl.pathname)) {
      if (isAuth && token.role) {
        redirectBasedOnRole(token.role as UserRole, pathname);
      }
    }

    if (isAuth) {
      if (token.isOnboarded) {
        const url = redirectBasedOnRole(token.role as UserRole, pathname);
        if (url && pathname === "/") {
          return NextResponse.redirect(new URL(url, req.url));
        }
      } else {
        if (pathname !== "/onboarding" && !isApiRoute) {
          return NextResponse.redirect(new URL("/onboarding", req.url));
        }
      }

      const role: string = token?.role as unknown as string;
      const routes = `/${role.toLowerCase().replace("_", "-")}`;
      const protectedRoutes = ["/admin", "/user"];

      if (protectedRoutes.some((path) => path.startsWith(path))) {
        if (isAuth && protectedRoutes.includes(pathname)) {
          if (pathname === routes || pathname.startsWith(routes)) {
            return;
          } else {
            return NextResponse.redirect(new URL(routes, req.url));
          }
        }
      }

      if (pathname === "/login" || pathname === "/register") {
        return NextResponse.redirect(new URL("/", req.url));
      }
    }
  },
  {
    callbacks: {
      async authorized() {
        await Promise.resolve();
        return true;
      },
    },
  },
);

export const config = {
  matcher: [
    "/((?!api|_next/static|_next/image|favicon.ico).*)",
    "/app/:path*",
    "/login",
    "/admin",
    "/user",
    "/register",
    "/api/:path*",
  ],
};
`

In the NextAuthOptions File

`ts
import { PrismaAdapter } from "@next-auth/prisma-adapter";
import { getServerSession, type NextAuthOptions } from "next-auth";
import CredentialsProvider from "next-auth/providers/credentials";
import bcrypt from "bcrypt";
import { type Session } from "next-auth";

import { db } from "~/server/db";

export const authOptions: NextAuthOptions = {
  secret: process.env.NEXTAUTH_SECRET,
  adapter: PrismaAdapter(db),
  session: {
    strategy: "jwt",
    maxAge: 60 * 60 * 24 * 30,
  },
  callbacks: {
    // eslint-disable-next-line
    async session({ token, session }) {
      if (token) {
        session.user.id = token.id;
        session.user.email = token.email;
        session.user.role = token.role;
        session.user.isOnboarded = token.isOnboarded;
      }
      return session;
    },
    // eslint-disable-next-line
    async jwt({ token, user, trigger, session: nextSession }) {
      const session = nextSession as Session;
      const dbUser = await db.user.findFirst({
        where: {
          email: token.email!,
        },
      });
      if (!dbUser) {
        token.id = user.id;
        return token;
      }
      if (trigger === "update") {
        return { ...token, ...session };
      }

      // @typescript-eslint/no-unnecessary-type-assertion
      return {
        id: dbUser.userId,
        email: dbUser.email,
        role: dbUser.role,
        isOnboarded: dbUser.isOnboarded,
      };
    },
  },
  providers: [
    CredentialsProvider({
      name: "credentials",
      credentials: {
        email: {
          type: "email",
        },
        password: {
          type: "password",
        },
      },
      async authorize(credentials) {
        if (!credentials) {
          return null;
        }
        if (!credentials.email || !credentials.password) {
          return null;
        }
        try {
          const user = await db.user.findFirst({
            where: {
              email: credentials.email,
            },
          });
          console.log(user);
          if (!user) {
            return null;
          }
          const isValidPassword = await bcrypt.compare(
            credentials.password,
            user.password,
          );
          if (!isValidPassword) {
            return null;
          }
          return {
            id: user.userId,
            email: user.email,
            role: user.role,
          };
        } catch (error) {
          return null;
        }
      },
    }),
  ],
  pages: {
    signIn: "/login",
  },
};

export const getServerAuthSession = () => getServerSession(authOptions);
`

Hope this helps ✌🏻
